livingatlas/pipelines/pom.xml

```diff
@@ -0,0 +1,890 @@
+                  <pattern>com.fasterxml.jackson</pattern>
+                  <shadedPattern>jjjj213.com.fasterxml.jackson</shadedPattern>
+                </relocation>
+                <relocation>
+                  <pattern>okhttp3</pattern>
+                  <shadedPattern>o115.okhttp3</shadedPattern>
+                </relocation>
+                <relocation>
+                  <pattern>okio</pattern>
+                  <shadedPattern>o113.okio</shadedPattern>
+                </relocation>
+                <relocation>
+                  <pattern>retrofit2</pattern>
+                  <shadedPattern>r240.retrofit2</shadedPattern>
+                </relocation>
+                <!-- Transient from elastic search Beam: conflicts with hive-jdbc-1.1.0-cdh5.12.1-standalone.jar -->
+                <relocation>
+                  <pattern>org.apache.http</pattern>
+                  <shadedPattern>hc45.org.apache.http</shadedPattern>
+                </relocation>
+                <!-- Transient from core: conflicts with Hadoop on Spark -->
+                <relocation>
+                  <pattern>org.hsqldb</pattern>
+                  <shadedPattern>h228.org.hsqldb</shadedPattern>
+                </relocation>
+                <relocation>
+                  <pattern>com.google.common</pattern>
+                  <shadedPattern>g20.com.google.common</shadedPattern>
+                </relocation>
+              </relocations>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>io.fabric8</groupId>
+        <artifactId>docker-maven-plugin</artifactId>
+        <version>${io.fabric8.version}</version>
+        <extensions>true</extensions>
+        <executions>
+          <execution>
+            <id>start</id>
+            <phase>pre-integration-test</phase>
+            <goals>
+              <goal>start</goal>
+            </goals>
+          </execution>
+          <execution>
+            <id>stop</id>
+            <phase>post-integration-test</phase>
+            <goals>
+              <goal>stop</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <images>
+            <image>
+              <alias>solr8</alias>
+              <name>djtfmartin/biocache-pipelines-solr8:v3</name>
+              <external>
+                <type>compose</type>
+                <composeFile>solr8.yml</composeFile>
+              </external>
+            </image>
+            <image>
+              <alias>ala-nameservice</alias>
+              <name>djtfmartin/ala-namematching-service:v20200214-4</name>
+              <external>
+                <type>compose</type>
+                <composeFile>ala-nameservice.yml</composeFile>
+              </external>
+              <run>
+                <wait>
+                  <http>
+                    <url>http://localhost:9179/api/search?q=Acacia</url>
```

Comment:
This looks suspicious - what if the port is in use by another test running on Jenkins? Should it be port scanning and assigning ports dynamically?

Reply:
No idea how to achieve that.

Reply:
We can look. In jenkins you can do things like this, but the app needs to allow you provide a port:

<img width="938" alt="Screenshot 2020-07-16 17 17 14" src="https://user-images.githubusercontent.com/237221/87689297-55f7bc00-c788-11ea-9d44-aa813239cb10.png">

In code, we do things [like this](https://github.com/gbif/registry/blob/3b7b97c2f7dc369de7d465753f41b6447d8b0415/registry-cli/src/test/java/org/gbif/registry/cli/datasetindex/DatasetBatchIndexerIT.java#L68)

Can you please see if it's possible to pass in a port at startup? It's a nuisance doing a release and 20 minutes in you get a port collision that requires manual intervention to rollback. If not, we should open an issue for someone to investigate this.

Reply:
Please share a bit more of this example in the screenshot, including the actual POM in use (a snippet if the not the whole) and we'll try and copy this pattern.

Reply:
i should add that I don't currently have access to the environment to test this.

Reply:
Can we run 'docker-compose down' to stop all running containers before we run tests?

Reply:
It's the registry project: https://github.com/gbif/registry/blob/master/pom.xml

I think `jetty.port` is probably used automatically by Jetty, I don't see any reference to it in our code. `System.getProperty`, anyway.

Reply:
The build command in Jenkins is now `mvn clean install verify dependency:analyze $SONAR_MAVEN_GOAL -Dhttp.port=$HTTP_PORT -Dsonar.host.url=$SONAR_HOST_URL -Dsonar.login=$SONAR_AUTH_TOKEN -U -P coverage`.

Reply:
Please bear in mind I only looked quickly.

The examples I listed were to show that Jetty can assign TCP PORT randomly per execution as a system variable. Here I think this relates to port mapping for docker containers, which are defined in the YAML? That may become much more convoluted (e.g. templating the prop files) unless the maven plugin lets us provide that from system variables.

If it is convoluted, I suggest we open an issue to "Parameterize the port in docker tests to allow concurrent executions" and then comment out their execution with a reference to the issue to let this progress.

Sound reasonable?

Reply:
Sorry - do you mean comment out the execution of the docker containers ?

Reply:
If it is the case that concurrent builds in jenkins all use the same ports in their tests, then yes (i.e. disable the tests until we solve that). If Jenkins doesn't run these, it can be ignored completely.

Reply:
Fair enough. It needs to be solved at some point. 
It was only if we cared to delay it to not block this merge.

Reply:
So to confirm, this needs to fixed before a merge is accepted ?

Reply:
https://github.com/gbif/pipelines/issues/337 I'll look into this, but others are welcome too

Reply:
this is resolved with the work on #337.

@qifeng-bai the maven build uses different ports for the docker containers to the ports specified in the docker-compose files described in the "Getting Started" section. This separation should help avoid the ports classh with maven builds from dev work in IDEs.